### PR TITLE
Package optimizer binary when releasing new version 

### PR DESCRIPTION
**Pull request by sctb512**
As titled.

### CODE REVIEW

**Comment by sctb512** [comment]
A test for this PR:
https://github.com/sctb512/nydus-snapshotter/releases/tag/v0.8.0

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -25,17 +25,48 @@ jobs:
...
+            ${{ runner.os }}-cargo
       - name: build nydus-snapshotter
         run: |
           go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.51.2
           export PATH=$PATH:$(go env GOPATH)/bin
           make
+      - name: build optimizer
+        run: |
+          rustup component add rustfmt clippy
+          make optimizer
```
Comment (changweige) [inline .github/workflows/release.yml:48]: It's better to release static binaries of optimizer

Comment (sctb512) [reply .github/workflows/release.yml:48]: Got it.
I also add a package `nydus-snapshotter-static-vX.X.X-x86_64.tgz`, which contains all static binaries. So, can we keep binaries in `nydus-snapshotter-vX.X.X-x86_64.tgz` is not static?

**Comment by imeoer** [comment]
image

**Comment by sctb512** [comment]
> What's the difference between static and non-static versions?

Here is the difference:
Static version:
```shell
❯ ldd nydus-snapshotter-static/*
nydus-snapshotter-static/02-optimizer-nri-plugin:
	not a dynamic executable
nydus-snapshotter-static/containerd-nydus-grpc:
	not a dynamic executable
nydus-snapshotter-static/optimizer-server:
	statically linked
```
Non-static version:
```shell
❯ ldd nydus-snapshotter/*
nydus-snapshotter/02-optimizer-nri-plugin:
	linux-vdso.so.1 (0x00007ffd38990000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f8077769000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f80775a9000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f8077796000)
nydus-snapshotter/containerd-nydus-grpc:
	linux-vdso.so.1 (0x00007ffffbd11000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fa2bd053000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fa2bce93000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fa2bd080000)
nydus-snapshotter/optimizer-server:
nydus-snapshotter/optimizer-server: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by nydus-snapshotter/optimizer-server)
nydus-snapshotter/optimizer-server: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by nydus-snapshotter/optimizer-server)
nydus-snapshotter/optimizer-server: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by nydus-snapshotter/optimizer-server)
	linux-vdso.so.1 (0x00007ffc69dd5000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f5e193bc000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f5e191fc000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f5e19451000)
```

File: .github/workflows/release.yml (hunk 1 above)
Comment (sctb512) [reply .github/workflows/release.yml:48]: In my machine, I got an error:
```
nydus-snapshotter/optimizer-server: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by nydus-snapshotter/optimizer-server)
nydus-snapshotter/optimizer-server: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by nydus-snapshotter/optimizer-server)
nydus-snapshotter/optimizer-server: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by nydus-snapshotter/optimizer-server)
```

It is better to release static binary for optimizer-server.

**Comment by sctb512** [comment]
Now, optimizer-server is always a static version.
```shell
❯ ldd nydus-snapshotter/*
nydus-snapshotter/02-optimizer-nri-plugin:
	linux-vdso.so.1 (0x00007ffc4f3ed000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f030f045000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f030ee85000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f030f072000)
nydus-snapshotter/containerd-nydus-grpc:
	linux-vdso.so.1 (0x00007ffc973e7000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f4c17120000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f4c16f60000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f4c1714d000)
nydus-snapshotter/optimizer-server:
	statically linked
❯ ldd nydus-snapshotter-static/*
nydus-snapshotter-static/02-optimizer-nri-plugin:
	not a dynamic executable
nydus-snapshotter-static/containerd-nydus-grpc:
	not a dynamic executable
nydus-snapshotter-static/optimizer-server:
	statically linked
```

**Comment by imeoer** [comment]
@sctb512 okay, why do we need a non-static version in the release?

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -25,17 +25,49 @@ jobs:
...
+          make static-optimizer-server
       - name: upload artifacts
         uses: actions/upload-artifact@v3
         with:
           name: nydus-snapshotter_artifacts
           path: |
             bin/containerd-nydus-grpc
+            bin/02-optimizer-nri-plugin
```
Comment (imeoer) [inline .github/workflows/release.yml:56]: Is it any special to add the prefix `02-`?

**Comment by sctb512** [comment]
> @sctb512 okay, why do we need a non-static version in the release?

We have static version binaries. Non-static version binaries, which have a smaller file size, are also required. Libraries used by non-static binaries can be upgraded dynamically.

File: .github/workflows/release.yml (hunk 2 above)
Comment (sctb512) [reply .github/workflows/release.yml:56]: > Is it any special to add the prefix `02-`?

We can not load the NRI plugin without the NRI plugin index `02-`. I don't know if this is a bug, but we can add this prefix when installing `optimizer-nri-plugin`.

**Comment by imeoer** [comment]
> > @sctb512 okay, why do we need a non-static version in the release?
> 
> We have static version binaries. Non-static version binaries, which have a smaller file size, are also required. Libraries used by non-static binaries can be upgraded dynamically.

Is it necessary to support the library upgrade? Most releases don't seem to have this requirement cc @changweige

**Comment by sctb512** [comment]
> Most releases don't seem to have this requirement

Got it. I 'll take it off if this is not necessary.

I find that `containerd` have released static and non-static versions.
By the way, our release packages do not specify architecture. Are they compatible with all platforms?

**Comment by changweige** [comment]
I think only releasing static binaries including `containerd-nydus-grpc` and other tools should meet our requirements now. We can also release other platform like `arm64`, however, it is low-priority now.

**Comment by sctb512** [comment]
Updated.
https://github.com/sctb512/nydus-snapshotter/releases/tag/v0.8.1

```shell
❯ ldd nydus-snapshotter/*
nydus-snapshotter/containerd-nydus-grpc:
	not a dynamic executable
nydus-snapshotter/optimizer-nri-plugin:
	not a dynamic executable
nydus-snapshotter/optimizer-server:
	statically linked
```

PTAL @changweige @imeoer

**Comment by imeoer** [comment]
Thanks, indeed containerd provided non-static versions, if snapshotter also does that, nydusd should do the same.

> By the way, our release packages do not specify architecture. Are they compatible with all platforms?

Nydusd [supports](https://github.com/dragonflyoss/image-service/releases/tag/v2.1.5) more platforms, we can do it later.

**Review by imeoer — APPROVED** [review]

**Review by changweige — APPROVED** [review]

**Pull request merged**